Save and load all preimages on the database

linked0 opened the pull request:
A validator must periodically reveal the preimages.
For that requirement, the preimages should be stored to database for a cycle.

This implementation is the first definition of done of the issue #492

linked0 commented:
Ready for review

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -33,9 +33,47 @@ import d2sqlite3.sqlite3;
+public struct PreimageStore
+{
+    /// Preimages of hashes of random value
+    public Hash[] data;
```
We could avoid writing all this code if we added serialization support to `Hash[]` in `Serializer.d` and `Deserializer.d`.

@Geod24 can you think of any drawbacks?

linked0 replied:
After adding serialization support to `Hash[]` in this PR, I will request a review again.

Geod24 replied:
@AndrejMitrovic : Actually for the binary protocol, I wrote a way to serialize ranges... Guess it's time to PR it.

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -104,6 +107,17 @@ public class EnrollmentManager
+        results = this.db.execute("SELECT val FROM node_enroll_data " ~
+            "WHERE key = ?", "preimages");
+
+        foreach (row; results)
```
I think maybe this would work too:

```D
if (!results.empty)
    this.preimages = results.oneValue!(ubyte[]).deserializeFull!(Hash[]);
```

Could be worth trying.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -319,10 +333,13 @@ public class EnrollmentManager
         foreach (i; 0 .. this.data.cycle_length-1)
-            this.data.random_seed = hashFull(this.data.random_seed);
+            this.preimages ~= hashFull(this.preimages[i]);
+        this.data.random_seed = this.preimages[this.data.cycle_length-1];
```
Why not just `this.preimages[$ - 1];`? Is it not the same?

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -453,6 +506,38 @@ public class EnrollmentManager
     }
 }
 
+/// test for serialization and deserialization of PreimageStore
```
Slightly outdated comment, as `PreimageStore` doesn't exist anymore.

AndrejMitrovic reviewed (changes requested):
small issues

linked0 commented:
Ready for review. Thanks.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -453,6 +502,38 @@ public class EnrollmentManager
 }
 
+/// test for serialization and deserialization of a Hash array
+unittest
```
This seems to just test serialization? It's not testing EnrollmentManager.

linked0 replied:
No, It's not. I will make the unittest with EnrollmentManager.

linked0 committed:
Save and load all pre-images on the database

A validator must periodically reveal the pre-images.
For that requirement, the pre-images should be stored to database for a cycle.

linked0 committed:
Add serialization/deserialization support to Hash[]

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -587,4 +672,13 @@ unittest
     assert(enrolls.isStrictlyMonotonic!("a.utxo_key < b.utxo_key"));
+
+    // test serialization/deserializetion for pre-images
+    auto preimages_1 = man.getPreimages();
```
You can access the field directly, because private allows access from the same module.

So just `auto preimages_1 = man.preimages;`.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -587,4 +672,13 @@ unittest
+    auto preimages_1 = man.getPreimages();
+    auto preimages_2 = man.loadPreimages();
+    int true_count;
+    foreach (i; 0 .. 1008)
```
These few lines can be replaced with `assert(preimages_1[] == preimages_2[]);`. This will check the length too.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -587,4 +672,13 @@ unittest
+    foreach (i; 0 .. 1008)
+        if (preimages_1[i] == preimages_2[i])
+            true_count++;
+    assert(true_count == 1008);
```
There is a missing check that the preimages are actually preimages. You can use this:

```D
import std.algorithm;
assert(equal!((a, b) => a.hashFull() == b)(preimages_1[0 .. $ - 1], preimages_1[1 .. $]));
```

linked0 commented:
Ready for review after applying the suggestions of @AndrejMitrovic .

AndrejMitrovic reviewed (approved):
LGTM.

linked0 commented:
Ping @Geod24
